http-response/background.js

```diff
@@ -1,22 +1,154 @@
 function listener(details) {
+  // The received data is a stream of bytes. In order to do text-based
+  // modifications, it is necessary to decode the bytes into a string
+  // using the proper character encoding, do any modifications, then
+  // encode back into a stream of bytes.
+  // Historically, detecting character encoding has been a tricky task
+  // taken on by the browser. Here, a simplified approach is taken
+  // and the complexity is hidden in a helper method.
+  let decoder, encoder;

```

Comment:
Keep the `TextDecoder` and `TextEncoder` constructors here as before to improve readability. You can declare a `charset` variable if you want to use something else instead of UTF-8.

Reply:
Well, except that I need to potentially modify the Content-Type output to be UTF-8 if the input type is not already; that would drive having something like `detectCharset` and `forceCharset` as separate calls. Would you find that organization to be preferable to the problematic `detectCharsetAndSetupDecoderEncoder` above?

Reply:
In the simplest case where the Content-Type needs to be an exact match for something that looks like `text/html; charset=utf-8", then you don't need to change the response header. But even if you do change the header, then you can still improve the readability by keeping the main logic in the listener, and using small helper functions to avoid too large function bodies. For example:

```javascript
function listener(details) {
  let {responseHeaders} = details;

  let ctHeader = responseHeaders.find(h => h.name.toLowerCase() == "content-type");
  // TODO: Verify that ctHeader is supported.
  // If Content-Type header not set, the browser is going to do content-sniffing,
  // and we should also return to avoid trouble (e.g. breaking downloads, PDFs, videos, ...).
  if (content type is not supported) {
    return;
  }

  let charset = ...
  let decoder = new TextDecoder(charset);
  let encoder = new TextEncoder();
  etc...
}
```

Reply:
I took another run at the code. I was able to clean it up in a few places as per your comments - keep the header object reference rather than index, return earlier/don't guess, etc. Still, a full *half* of the code is dedicated to character encoding alone; I can't say that the code is elegant.